src/primitives.rs

```diff
@@ -0,0 +1,72 @@
+use crate::{WideWord, Word};
+
+/// Multiplies `x` and `y`, returning the most significant
+/// and the least significant words as `(hi, lo)`.
+#[inline(always)]
+pub(crate) const fn mulhilo(x: Word, y: Word) -> (Word, Word) {
+    let res = (x as WideWord) * (y as WideWord);
+    ((res >> Word::BITS) as Word, res as Word)
+}
+
+/// Adds wide numbers represented by pairs of (most significant word, least significant word)
+/// and returns the result in the same format `(hi, lo)`.
+#[inline(always)]
+pub(crate) const fn addhilo(x_hi: Word, x_lo: Word, y_hi: Word, y_lo: Word) -> (Word, Word) {
+    let res = (((x_hi as WideWord) << Word::BITS) | (x_lo as WideWord))
+        + (((y_hi as WideWord) << Word::BITS) | (y_lo as WideWord));
+    ((res >> Word::BITS) as Word, res as Word)
+}
+
+/// Computes `lhs + rhs + carry`, returning the result along with the new carry (0, 1, or 2).
+#[inline(always)]
+pub const fn adc3(lhs: Word, rhs: Word, carry: Word) -> (Word, Word) {
+    let a = lhs as WideWord;
+    let b = rhs as WideWord;
+    let carry = carry as WideWord;
+    let ret = a + b + carry;
+    (ret as Word, (ret >> Word::BITS) as Word)
+}
+
+/// Computes `lhs + rhs`, returning the result along with the carry (0 or 1).
+#[inline(always)]
+pub const fn adc(lhs: Word, rhs: Word) -> (Word, Word) {
+    let (res, carry) = lhs.overflowing_add(rhs);
```

Comment:
This should be constant-time according to Godbolt. Safe to use?

Reply:
The main thing that seems weird is that it doesn't accept a carry as input, so you can't actually use it to build a carry chain?

Reply:
I guess that's what you're calling `adc3` and I suppose they're both valuable

Reply:
Yes, judging by the benchmarks the compiler doesn't optimize out `carry = 0`.

Reply:
Aah cool, I like this being renamed to `overflowing_add`, but in that case perhaps we can just use the method on the primitive type?

Reply:
I wanted it to imply that using `Word::overflowing_add()` was "vetted" as opposed to just using it directly which may seem like an omission. Also make it easier to try different approaches e.g. via `WideWord`.